File: solution/912. Sort an Array/912_02.cpp
class Solution {
public:
    vector<int> sortArray(vector<int>& nums) {
        mergeSort(nums, 0, nums.size() - 1);
        return nums;
    }
    
private:
    void mergeSort(vector<int>& nums, int l, int r) {
        if (l >= r) return;
        
        int m = l + (r - l) / 2;
        mergeSort(nums, l, m);
        mergeSort(nums, m + 1, r);
        
        int len = r - l + 1;
        vector<int> sorted(len);
        int i = l, j = m + 1;
        int k = 0;
        while (i <= m && j <= r) {
            sorted[k++] = nums[i] < nums[j] ? nums[i++] : nums[j++];
        }
        while (i <= m) {
            sorted[k++] = nums[i++];
        }
        while (j <= r) {
            sorted[k++] = nums[j++];
        }
        for (int k = 0; k < len; ++k)
            nums[k + l] = sorted[k];
    }
};

File: solution/912. Sort an Array/912_02.py
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        def merge_sort(A, l, r):
            if l >= r: return
            
            m = (l + r) // 2
            merge_sort(A, l, m)
            merge_sort(A, m + 1, r)
            
            n = r - l + 1
            B = [0] * n
            
            i, j = l, m + 1
            k = 0
            
            while i <= m and j <= r:
                if A[i] < A[j]:
                    B[k] = A[i]
                    i += 1
                else:
                    B[k] = A[j]
                    j += 1
                k += 1
                
            while i <= m:
                B[k] = A[i]
                i += 1
                k += 1
                
            while j <= r:
                B[k] = A[j]
                j += 1
                k += 1
            
            for k in range(n):
                A[l + k] = B[k]
            
        merge_sort(nums, 0, len(nums) - 1)
        return nums
    
